controlplane/kubeadm/controllers/controller.go

```diff
@@ -588,3 +600,54 @@ func (r *KubeadmControlPlaneReconciler) adoptOwnedSecrets(ctx context.Context, k
 
 	return nil
 }
+
+func (r *KubeadmControlPlaneReconciler) remediateUnhealthy(ctx context.Context, logger logr.Logger, controlPlane *internal.ControlPlane) error {
+	// No remediation while cluster is below remediation threshold, or if the cluster is scaling up
+	// This is to prevent multiple remediations from happening before replacements are created.
+	if !controlPlane.RemediationAllowed() || !controlPlane.HasUnhealthyMachine() || controlPlane.Machines.Len() < int(*controlPlane.KCP.Spec.Replicas) {
+		return nil
+	}
+
+	if controlPlane.HasDeletingMachine() {
+		return &capierrors.RequeueAfterError{RequeueAfter: deleteRequeueAfter}
+	}
+
+	workloadCluster, err := r.managementCluster.GetWorkloadCluster(ctx, util.ObjectKey(controlPlane.Cluster))
+	if err != nil {
+		return errors.Wrap(err, "failed to create workload cluster client")
+	}
+
+	etcdStatus, err := workloadCluster.EtcdStatus(ctx)
+	if err != nil {
+		return errors.Wrap(err, "failed to fetch etcd status")
+	}
+
+	machine := controlPlane.UnhealthyMachines().Oldest()
```

Comment:
Should we consider the case when there are more than one unhealty machine. e.g. two unhealty, and the first one can't be remediated due to etcd quota, but the second one yes

Reply:
Yeah this occurred to me, but I think we can leave it out for now and revisit this later if it comes up